// http://www.postgresql.org/docs/9.3/static/functions-info.html#FUNCTIONS-TXID-SNAPSHOT-PARTS
message Snapshot {

    required uint64 min = 1;

    required uint64 max = 2;

    repeated uint64 active = 3;

}

// https://github.com/google/protobuf/blob/master/src/google/protobuf/timestamp.proto#L95-L106
message Timestamp {

  required int64 seconds = 1;

  required int32 nanos = 2;

}


message BatchIdentifier {

    required uint64 id = 1;

    required bytes node = 2;

}

// https://github.com/markokr/skytools/blob/master/sql/pgq/structure/tables.sql#L102-L124
message Tick {

    required uint64 id = 1;

    required Snapshot snapshot = 2;

    required Timestamp timestamp = 3;

}


message Column {

    // Should be unique.
    required string name = 1;

    // Not required. (Values that are not present are `NULL`/`None`.)
    // This follows the same conversion rules as PL/Python:
    // http://www.postgresql.org/docs/9.3/static/plpython-data.html
    oneof value {

        bool boolean = 2;
        int64 integer64 = 3;
        float float = 4;

        // TODO: This needs to ensure that strings are transcoded from
        // the server encoding to UTF-8.
        string string = 5;  // UTF-8

    }

}


message Row {

    repeated Column columns = 1;

}


message Message {

    message Header {

        // Unique identifier of the process that sent this message.
        required bytes publisher = 1;

        // A serial number for each message sent by this publisher. This should
        // be increased by one for each message sent/seen. Allows ensuring all
        // messages are seen by the consumer in order, without gaps.
        required uint64 sequence = 2;

        // When this message was generated. This is solely for debugging and
        // monitoring purposes, such as identifying how far behind the replication
        // stream is in seconds rather than bytes.
        required Timestamp timestamp = 3;

    }

    required Header header = 1;

    // Required.
    oneof operation {

        BatchOperation batch_operation = 2;

    }

}


message BatchOperation {

    required BatchIdentifier batch_identifier = 1;

    // Required
    oneof operation {
        BeginOperation begin_operation = 2;
        MutationOperation mutation_operation = 3;
        CommitOperation commit_operation = 4;
        RollbackOperation rollback_operation = 5;
    }

}

message BeginOperation {

    required Tick start = 1;

    required Tick end = 2;

}


message MutationOperation {

    enum Operation {
        INSERT = 1;
        UPDATE = 2;
        DELETE = 3;
    }

    required uint64 id = 1;

    required string schema = 2;

    required string table = 3;

    required Operation operation = 4;

    // A column names that can be used to uniquely identify this row in it's
    // table. Generally, this is the primary key(s) of the row, but could also
    // be a unique (and not null) index.
    repeated string identity_columns = 5;

    // Must not exist for INSERT, but must exist for UPDATE and DELETE.
    optional Row old = 6;

    // Must not exist for DELETE, but must exist for INSERT and UPDATE.
    optional Row new = 7;

    required Timestamp timestamp = 8;

    required uint64 transaction = 9;

}


message CommitOperation {}


message RollbackOperation {}
